Local redeployment

yesterday I got very bored by restarting my Application Server (WildFly) every time when I assemble my EAR and using `cargoStartLocal`/`cargoRunLocal`, because WildFly is able to redeploy it on runtime.

## Primary Issue: ##
So what I am missing, is a task `cargoRedeployLocal` on a running instance. For convenience I would also suggest a task cargoDeployLocal, which dependsOn `cargoStartLocal`, if not yet started. So this local tasks are my primary issue.



## Workaround 1 (without success): ##
The suggestion in https://github.com/bmuschko/gradle-cargo-plugin/issues/102 of using cargo*Local for start/stop and redeploy with cargoRedeployRemote did not work for me. Perhaps we could make this work. Here is my config for cargo:

```gradle
cargo {
    containerId = 'wildfly8x'
    deployable {
        file = file('myApplication.ear')
    }
    ...
    remote {
        hostname = 'localhost'
        // this should correspond to the user in /config/wildfly/mgmt-users.properties, your local wildfly admin.
        username = 'admin'
        password = 'admin'
        
        // containerProperties {
        //     property 'cargo.jboss.management-http.port', 8787 // default: 9990
        // }
        // containerProperties {
        //     property 'cargo.protocol', "http" // ?!? does not make any difference on cargoRedeployRemote!
        // }
        // containerProperties {
        //     property 'cargo.timeout', "60000" // ?!? does not make any difference on cargoRedeployRemote!
        // }
    }
    local {
        ...
        jvmArgs = '-Xdebug -Xrunjdwp:transport=dt_socket,address=8787,server=y,suspend=n'
        installer {
            installUrl = 'http://download.jboss.org/wildfly/8.2.0.Final/wildfly-8.2.0.Final.tar.gz'
            downloadDir = file("$wildfly/download")
            extractDir = file("$wildfly")
        }
        ...
    }
}
```

With gradle cargoRedeployRemote --stacktrace this results in:
```
...
Caused by: java.net.ConnectException: JBAS012144: Could not connect to remote://localhost:9990. The connection timed out
    at org.jboss.as.protocol.ProtocolConnectionUtils.connectSync(ProtocolConnectionUtils.java:130)
```

In my standalone.xml this port 9990 is defined here:
```xml
<server>
    ...
    <socket-binding-group name="standard-sockets" default-interface="public" port-offset="${jboss.socket.binding.port-offset:0}">
        <socket-binding name="management-http" interface="management" port="${jboss.management.http.port:9990}"/>
        ...
        </socket-binding-group>
</server>
```
In the WildFly Console (the browser-GUI) you can also check under Configuration-->Socket Binding that "management-http" is an available socket. So, perhaps I have to define another port instead of "management-http" or something. Any help would be appreciated.



## Workaround 2: ##
My current workaround is a gradle copy task:
```gradle
task redeploy (type: Copy) {
	from file("myApplication.ear")
	into System.properties['user.home'] + "\\AppData\\Local\\Temp\\cargo\\conf\\deployments"
}
```
On file base you can see that WildFly will recognize the new *.ear and start deploying (creating files "myApplication.ear.isdeploying" and "myApplication.ear.deployed").
But IMHO this should be part of the gradle-cargo-plugin, because the locations (from, into) are already known by the plugin.

--- a/.gitignore
+++ b/.gitignore
@@ -5,3 +5,7 @@ out
 *.iws
 .idea
 .gradle
+/bin/
+/.project
+/.classpath
+/.settings

--- a/README.md
+++ b/README.md
@@ -126,6 +126,12 @@ The `cargo` plugin pre-defines the following tasks out-of-the-box:
         <td>Starts the local container, deploys web application to it and then do other tasks (for example, execute tests).</td>
     </tr>
     <tr>
+        <td>cargoRedeployLocal</td>
+        <td>-</td>
+        <td><a href="http://bmuschko.github.io/gradle-cargo-plugin/docs/groovydoc/com/bmuschko/gradle/cargo/tasks/local/CargoRedeployLocal.html">CargoRedeployLocal</a></td>
+        <td>Undeploy and deploy again a deployable. If the deployable was not deployed before calling redeploy, it will simply be deployed.</td>
+    </tr>
+    <tr>
         <td>cargoStopLocal</td>
         <td>-</td>
         <td><a href="http://bmuschko.github.io/gradle-cargo-plugin/docs/groovydoc/com/bmuschko/gradle/cargo/tasks/local/CargoStopLocal.html">CargoStopLocal</a></td>

--- a/build.gradle
+++ b/build.gradle
@@ -19,7 +19,7 @@ def compatibilityVersion = 1.6
 sourceCompatibility = compatibilityVersion
 targetCompatibility = compatibilityVersion
 group = 'com.bmuschko'
-version = '2.1'
+version = '2.2'
 
 repositories {
     mavenCentral()

--- a/src/main/groovy/com/bmuschko/gradle/cargo/CargoPlugin.groovy
+++ b/src/main/groovy/com/bmuschko/gradle/cargo/CargoPlugin.groovy
@@ -22,6 +22,7 @@ import com.bmuschko.gradle.cargo.property.CargoProjectProperty
 import com.bmuschko.gradle.cargo.property.LocalContainerTaskProperty
 import com.bmuschko.gradle.cargo.property.RemoteContainerTaskProperty
 import com.bmuschko.gradle.cargo.tasks.AbstractCargoContainerTask
+import com.bmuschko.gradle.cargo.tasks.local.CargoRedeployLocal
 import com.bmuschko.gradle.cargo.tasks.local.CargoRunLocal
 import com.bmuschko.gradle.cargo.tasks.local.CargoStartLocal
 import com.bmuschko.gradle.cargo.tasks.local.CargoStopLocal
@@ -134,6 +135,7 @@ class CargoPlugin implements Plugin<Project> {
         project.task('cargoRunLocal', type: CargoRunLocal)
         project.task('cargoStartLocal', type: CargoStartLocal)
         project.task('cargoStopLocal', type: CargoStopLocal)
+        project.task('cargoRedeployLocal', type: CargoRedeployLocal)
     }
 
     private void checkValidContainerId(Project project, CargoPluginExtension cargoPluginExtension) {

--- a/src/main/groovy/com/bmuschko/gradle/cargo/tasks/local/CargoRedeployLocal.groovy
+++ b/src/main/groovy/com/bmuschko/gradle/cargo/tasks/local/CargoRedeployLocal.groovy
@@ -1,0 +1,28 @@
+/*
+ * Copyright 2013 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.bmuschko.gradle.cargo.tasks.local
+
+/**
+ * Undeploy and deploy again a deployable to local. If the deployable was not deployed before calling redeploy, it will simply be deployed.
+ *
+ * @author Sebl29@users.noreply.github.com
+ */
+class CargoRedeployLocal extends LocalCargoContainerTask {
+    CargoRedeployLocal() {
+        action = 'redeploy'
+        description = 'Undeploy and deploy again a deployable. If the deployable was not deployed before calling redeploy, it will simply be deployed.'
+    }
+}

